File: src/main/java/core/procedures/hashmaps/Vals.java
package core.procedures.hashmaps;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMCons;

import java.util.Map;

public final class Vals extends AFn {

  public Vals() {
    super(new FnArgsBuilder().min(1).max(1).mandatory(new Class[]{Map.class}).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "vals";
  }

  @Override
  public Object apply1(Object arg) {
    return SCMCons.list(((Map)arg).values());
  }
}

File: src/main/java/core/procedures/hashmaps/Vals.kt
package core.procedures.hashmaps

import core.procedures.AFn
import core.procedures.Arity.Exactly

class Vals : AFn<Map<*, *>?, Sequence<*>>(name = "vals", isPure = true, arity = Exactly(1),
                                          mandatoryArgsTypes = arrayOf(Map::class.java)) {

    override operator fun invoke(arg: Map<*, *>?) = arg!!.values.asSequence()
}
